pkg/kp/consulutil/watch.go

```diff
@@ -0,0 +1,165 @@
+package consulutil
+
+import (
+	"time"
+
+	"github.com/square/p2/Godeps/_workspace/src/github.com/hashicorp/consul/api"
+)
+
+// WatchPrefix watches a Consul prefix for changes to any keys that have the prefix. When
+// anything changes, all Key/Value pairs having that prefix will be written to a provided
+// channel.
+//
+// Errors will sent on the given output channel but do not otherwise affect execution. The
+// provided output stream will become owned by this function call, and this call will
+// close it when the function ends. This function will run until explicitly canceled by
+// closing the "done" channel. Data is written to the output channel synchronously, so
+// readers must consume the data to this method will block.
+func WatchPrefix(
+	prefix string,
+	clientKV ConsulLister,
+	outPairs chan<- api.KVPairs,
+	done <-chan struct{},
+	outErrors chan<- error,
+) {
+	defer close(outPairs)
+	var currentIndex uint64
+	timer := time.NewTimer(time.Duration(0))
+
+	for {
+		select {
+		case <-done:
+			return
+		case <-timer.C:
+		}
+		timer.Reset(250 * time.Millisecond) // upper bound on request rate
+		pairs, queryMeta, err := SafeList(clientKV, done, prefix, &api.QueryOptions{
+			WaitIndex: currentIndex,
+		})
+		switch err {
+		case CanceledError:
+			return
+		case nil:
+			currentIndex = queryMeta.LastIndex
+			select {
+			case <-done:
+			case outPairs <- pairs:
+			}
+		default:
+			select {
+			case <-done:
+			case outErrors <- err:
+			}
+			timer.Reset(2 * time.Second) // backoff
+		}
+	}
+}
+
+type NewKeyHandler func(key string) chan<- *api.KVPair
+
+type keyMeta struct {
+	created    uint64
+	modified   uint64
+	subscriber chan<- *api.KVPair
+}
+
+// WatchNewKeys watches for changes to a list of Key/Value pairs and lets each key be
+// handled individually though a subscription-like interface.
+//
+// This function models a key's lifetime in the following way. When a key is first seen,
+// the given NewKeyHandler function will be run, which may return a channel. When the
+// key's value changes, new K/V updates are sent to the key's notification channel. When
+// the key is deleted, `nil` is sent. After being deleted or if the watcher is asked to
+// exit, a key's channel will be closed, to notify the receiver that no further updates
+// are coming.
+//
+// WatchNewKeys doesn't watch a prefix itself--the caller should arrange a suitable input
+// stream of K/V pairs, probably from WatchPrefix(). This function runs until the input
+// stream closes. Closing "done" will asynchronously cancel the watch and cause it to
+// eventually exit.
+func WatchNewKeys(pairsChan <-chan api.KVPairs, onNewKey NewKeyHandler, done <-chan struct{}) {
```

Comment:
Interesting, is this a generic farm API? I like it already. However speaking from past experience - it is valuable to be able to block on the termination of this function _and_ all its children. Not implementing that soundly was a mistake in my original design which made teardown of farms a bit hairy at program exit. Might want to work in a sync.WaitGroup here.